enhancement: Refactor login to use refresh tokens [CORE-2747]

joelvdavies opened the pull request:
This PR:

- Overhauls the login functions and all requests, adding a DAFNISession object instead which handles refresh tokens should a request fail due to authentication on a request by request basis (CORE-2747 & CORE-2954 & CORE-2921)
- The current session data for the CLI is now stored in a file named .dafni-cli in the user's home directory (avoiding lots of these files being created everywhere)
- Adds a timeout of 100 seconds to all requests (see note on https://requests.readthedocs.io/en/latest/user/quickstart/#timeouts - it should not effect downloads) (CORE-2953)
- Keycloak is now called on logout to invalidate the token (CORE-2927)

Closes #59, Closes #58

sarahbyrnie commented on dafni_cli/api/session.py:
```diff
@@ -0,0 +1,547 @@
+
+        return response
+
+    # TODO have same optional flags for each function
```
You've potentially done this now that you use the _authenticated_request method?

joelvdavies commented on login_test.py:
```diff
@@ -0,0 +1,5 @@
+
+
+session = DAFNISession()
+# session.logout()
```
Nope I meant to delete it 😄

joelvdavies commented on dafni_cli/datasets/dataset_metadata.py:
```diff
@@ -76,7 +77,7 @@ class DatasetMetadata:
     Methods:
         __init__(): DatasetMetadata constructor
-        set_attributes_from_dict(jwt (str), dataset (dict)): Sets the dataset metadata attributes from given client model dict
+        set_attributes_from_dict(session (DAFNISession), dataset (dict)): Sets the dataset metadata attributes from given client model dict
```
Yeah, personally I would rather just remove these sections – its only repeating what methods are defined below and hasn’t been touched for 2 years and its just another place to update the comments unless there is some reason we would want to keep them? These changes came from a bulk find & replace so I didn't really touch it myself.

sarahbyrnie replied:
I believe we follow this standard https://peps.python.org/pep-0257/ which includes listing public class methods in the docstring

joelvdavies replied:
I have updated it for this PR anyway though (although I haven't added any missing ones if there are any) - I can address this properly in another PR if we want to keep them or otherwise remove them.

joelvdavies replied:
Ah yes I see, the specific part seems to be
"The docstring for a class should summarize its behavior and list the public methods and instance variables. If the class is intended to be subclassed, and has an additional interface for subclasses, this interface should be listed separately (in the docstring). The class constructor should be documented in the docstring for its __init__ method. Individual methods should be documented by their own docstring."

I did find https://stackoverflow.com/questions/62444612/should-i-list-class-methods-in-the-class-docstring, I cant see any reference to it explicitly being like we are using here with full repeated descriptions and parameter definitions, the sample from datetime only lists the function names and ignores the arguments. It seems that most ignore it entirely. I must admit I don't see why its a standard since vscode will just show the docstring when you look at the methods, but then it was written in 2001.

joelvdavies commented on dafni_cli/model/model.py:
```diff
@@ -156,27 +157,27 @@ def set_attributes_from_dict(self, model_dict: dict):
         self.version_message = model_dict["version_message"]
 
-    def get_metadata(self, jwt_string: str):
+    def get_metadata(self, session: str):
```
No, I will remove it now.

joelvdavies commented:
Ok I think I have addressed the comments. I have also tested it works on windows again and removed the user_id and jwt parsing as it isn't necessary to display the user id. pyjwt is definitely much simpler if we need it again in the future.
